ledger/src/blockstore.rs

```diff
@@ -1608,6 +1748,187 @@ impl Blockstore {
+                    duplicate_shreds.push(PossibleDuplicateShred::ChainedMerkleRootConflict(
+                        shred.clone(),
+                        next_shred,
+                    ));
+                }
+                return false;
+            }
+        }
+
+        true
+    }
+
+    /// Returns true if there is no chaining conflict between
+    /// the `shred` and `merkle_root_meta` of the previous FEC set,
+    /// or if shreds from the previous set are yet to be received.
+    ///
+    /// Otherwise return false and add duplicate proof to
+    /// `duplicate_shreds`.
+    ///
+    /// This is intended to be used right after `shred`'s `merkle_root_meta`
+    /// has been created for the first time.
+    fn check_backwards_chained_merkle_root_consistency(
+        &self,
+        shred: &Shred,
+        just_inserted_shreds: &HashMap<ShredId, Shred>,
+        erasure_metas: &BTreeMap<ErasureSetId, WorkingEntry<ErasureMeta>>,
+        merkle_root_metas: &mut HashMap<ErasureSetId, WorkingEntry<MerkleRootMeta>>,
+        duplicate_shreds: &mut Vec<PossibleDuplicateShred>,
+    ) -> bool {
+        let slot = shred.slot();
+        let erasure_set = shred.erasure_set();
+        let fec_set_index = shred.fec_set_index();
+
+        if fec_set_index == 0 {
+            // Although the first fec set chains to the last fec set of the parent block,
+            // if this chain is incorrect we do not know which block is the duplicate until votes
+            // are received. We instead delay this check until the block reaches duplicate
+            // confirmation.
+            return true;
+        }
+
+        // If a shred from the previous fec set has already been inserted, check the chaining.
+        // Since we cannot compute the previous fec set index, we check the in memory map, otherwise
+        // check the previous key from blockstore to see if it is consecutive with our current set.
+        let Some(prev_erasure_set) = self
+            .previous_erasure_set(erasure_set, erasure_metas)
+            .expect("Expect database operations to succeed")
+        else {
+            // No shreds from the previous erasure batch have been received,
+            // so nothing to check. Once the previous erasure batch is received,
+            // we will verify this chain through the forward check above.
+            return true;
+        };
+        if let HashMapEntry::Vacant(entry) = merkle_root_metas.entry(prev_erasure_set) {
+            if let Some(meta) = self.merkle_root_meta(prev_erasure_set).unwrap() {
+                entry.insert(WorkingEntry::Clean(meta));
+            }
+        }
+        let Some(prev_merkle_root_meta_entry) = merkle_root_metas.get(&prev_erasure_set) else {
+            // This can happen if the previous erasure meta was not consistent with the shred
+            // received. The merkle root meta is only created if the shred is inserted,
+            // however the erasure meta always exists.
```

Comment:
I don't understand this;
When do we create and store erasure meta from the shred received, but do not store its merkle_root meta?
Can you plz point me to where that logic is in the code?

Reply:
I think I'm wrong here. 

If the shred passes the preliminary checks, an erasure meta is created:
https://github.com/anza-xyz/agave/blob/5eab6ae2b5aec6910bd5a00bfccc00614450bec7/ledger/src/blockstore.rs#L1248-L1255

We then perform the following checks:
1) https://github.com/anza-xyz/agave/blob/5eab6ae2b5aec6910bd5a00bfccc00614450bec7/ledger/src/blockstore.rs#L1258
2) https://github.com/anza-xyz/agave/blob/5eab6ae2b5aec6910bd5a00bfccc00614450bec7/ledger/src/blockstore.rs#L1307-L1309

I had assumed (1) could fail, but looking at the code this seems impossible.

We finally insert the merkle meta if both checks succeed:
https://github.com/anza-xyz/agave/blob/5eab6ae2b5aec6910bd5a00bfccc00614450bec7/ledger/src/blockstore.rs#L1311-L1318
I had previously guarded this because the merkle root meta has a reference to the first received shred and I didn't want to end up in scenarios where this reference is inconsistent with blockstore.

Reply:
although it is unlikely (2) could fail, but then we probably have bigger issues.
So I think fine to just `unwrap / expect` here.